kernel/src/main/java/org/kframework/parser/json/JsonParser.java

```diff
@@ -124,16 +125,16 @@ public static Definition toDefinition(JsonObject data) throws IOException {
 
         String mainModuleName = data.getString("mainModule");
         JsonArray mods = data.getJsonArray("modules");
-        Set<FlatModule> entryModules = new HashSet<>();
+        List<FlatModule> flatModules = new ArrayList<>();
         for (JsonObject m: mods.getValuesAs(JsonObject.class)) {
-            entryModules.add(toFlatModule(m));
+            flatModules.add(toFlatModule(m));
         }
 
-        Map<String,Module> koreModules = new HashMap<>();
-        FlatModule mainFlatMod = entryModules.stream().filter(mod -> mod.name().equals(mainModuleName)).findFirst().get();
-        Module mainMod = mainFlatMod.toModule(immutable(entryModules), JavaConverters.mapAsScalaMapConverter(koreModules).asScala(), Seq());
-
-        return new Definition(mainMod, immutable(new HashSet<>(koreModules.values())), toAtt(data.getJsonObject("att")));
+        scala.collection.Set<Module> koreModules = FlatModule.toModule(immutable(flatModules), Set());
+        return Constructors.Definition(
+                koreModules.find(x -> x.name().equals(mainModuleName))
+                        .getOrElse(() -> { throw new AssertionError("Could not find main module name: " + mainModuleName); }),
+                koreModules, toAtt(data.getJsonObject("att")));
```

Comment:
Will this assertion error print out the JSON file that was being parsed or other source location information? Is it possible to?

IIRC, assertion errors only produce stack dumps?

Reply:
Also, isn't this somehow exactly the same code on line 107 of KILtoKORE.java? Should it be factored out to a common place? Maybe in `outer.scala`?

Reply:
This class doesn't have access to the source file. It receives only the contents.
I've chosen to throw AssertionError because I don't expect this to be thrown unless something very wrong is happening.
I've updated the error message so it points out that it happened when loading from a JSON file.
The instance from KILtoKORE is only used in UnitTests.

The code does look similar, but I don't think it's worth extracting a function. Even less now that I changed the error message.
There is no special self-contained logic in there, just an object creation with a very specific context.